### PR TITLE
Remove unused packages

### PR DESCRIPTION
**Pull request by newhoggy**
Remove unused packages.  These have been identified with `-Wunused-packages` which is available from `ghc-8.10.x` onwards.

Removed packages include:

* array
* bech32
* cardano-crypto-praos
* cardano-ledger-alonzo
* cardano-ledger-shelley-ma
* cborg-json
* containers
* doctest
* groups
* mtl
* plutus-ledger-api
* plutus-tx
* prettyprinter
* small-steps
* tasty-hunit

It's possible that some like `plutus-ledger-api` and `plutus-tx` have been added intentionally for future work.  Please let me know if this is the case.  `-Wunused-packages` can be removed temporarily for specific projects to accomodate this.

Also moved common project configuration to a `common project-config` stanza to reduce duplication and ensure the same configuration is used across the cabal file and simplify future configuration editing.  The move allows for the specification of `-Wunused-packages`, which requires an `if` to be done from a single place per cabal file as well.

`base` constraints are also now specified in a single place near the tope of the `cabal` file.  I haven't made them all the same across all cabal files in case they're intentionally different, but we can make them all the same if we desire consistency.

The hope is by removing unnecessary dependencies, we can reduce cabal solver time especially for downstream projects where solving time can be up to 11 minutes.

### CODE REVIEW

**Comment by newhoggy** [comment]
These test failures have been observed intermittently, but I don't think they are related to the changes in this PR:

```
    Only valid CHAIN STS signals are generated:                                                         FAIL (48.51s)
      *** Failed! (after 94 tests):
      Exception while generating shrink-list:
        LEDGERS sigGen: [[UtxowFailure (UtxoFailure (OutputTooSmallUTxO [(Addr Testnet (KeyHashObj (KeyHash "7ea150efc18283df")) (StakeRefBase (KeyHashObj (KeyHash "a4b2a435a88d40e8"))),Coin 0),(Addr Testnet (KeyHashObj (KeyHash "6cc5fd90b56b9244")) (StakeRefBase (KeyHashObj (KeyHash "085f13811231b936"))),Coin 0),(Addr Testnet (ScriptHashObj (ScriptHash "82df1dde2a9bae2d")) (StakeRefBase (ScriptHashObj (ScriptHash "c0ac5b0256afafcb"))),Coin 0),(Addr Testnet (ScriptHashObj (ScriptHash "d30ead5aaafc7ae9")) (StakeRefBase (ScriptHashObj (ScriptHash "c0ac5b0256afafcb"))),Coin 0)]))]]
        CallStack (from HasCallStack):
          error, called at src/Test/Shelley/Spec/Ledger/Generator/Trace/Ledger.hs:183:15 in shelley-spec-ledger-test-0.1.0.0-inplace:Test.Shelley.Spec.Ledger.Generator.Trace.Ledger
      Exception thrown while showing test case:
        LEDGERS sigGen: [[UtxowFailure (UtxoFailure (OutputTooSmallUTxO [(Addr Testnet (KeyHashObj (KeyHash "7ea150efc18283df")) (StakeRefBase (KeyHashObj (KeyHash "a4b2a435a88d40e8"))),Coin 0),(Addr Testnet (KeyHashObj (KeyHash "6cc5fd90b56b9244")) (StakeRefBase (KeyHashObj (KeyHash "085f13811231b936"))),Coin 0),(Addr Testnet (ScriptHashObj (ScriptHash "82df1dde2a9bae2d")) (StakeRefBase (ScriptHashObj (ScriptHash "c0ac5b0256afafcb"))),Coin 0),(Addr Testnet (ScriptHashObj (ScriptHash "d30ead5aaafc7ae9")) (StakeRefBase (ScriptHashObj (ScriptHash "c0ac5b0256afafcb"))),Coin 0)]))]]
        CallStack (from HasCallStack):
          error, called at src/Test/Shelley/Spec/Ledger/Generator/Trace/Ledger.hs:183:15 in shelley-spec-ledger-test-0.1.0.0-inplace:Test.Shelley.Spec.Ledger.Generator.Trace.Ledger
      
      Use --quickcheck-replay=333481 to reproduce.
```

```
      total amount of Ada is preserved (Chain):                                                           FAIL (286.82s)
        *** Failed! (after 185 tests):
        Exception while generating shrink-list:
          LEDGERS sigGen: [[UtxowFailure (UtxoFailure (FeeTooSmallUTxO (Coin 5991682) (Coin (-19))))]]
          CallStack (from HasCallStack):
            error, called at src/Test/Shelley/Spec/Ledger/Generator/Trace/Ledger.hs:183:15 in shelley-spec-ledger-test-0.1.0.0-inplace:Test.Shelley.Spec.Ledger.Generator.Trace.Ledger
        Use --quickcheck-replay=564994 to reproduce.
```

```
    total amount of Ada is preserved (Chain):                                                           FAIL (463.13s)
      *** Failed! (after 274 tests):
      Exception while generating shrink-list:
        LEDGERS sigGen: [[UtxowFailure (UtxoFailure (OutputTooSmallUTxO [(Addr Testnet (KeyHashObj (KeyHash "0e4a87c2e1ba65a4")) (StakeRefBase (KeyHashObj (KeyHash "953f0b90bca642e6"))),Coin 1),(Addr Testnet (ScriptHashObj (ScriptHash "a8bf76435c74fec2")) (StakeRefBase (ScriptHashObj (ScriptHash "fc4e869e410d4a58"))),Coin 1)]))]]
        CallStack (from HasCallStack):
          error, called at src/Test/Shelley/Spec/Ledger/Generator/Trace/Ledger.hs:183:15 in shelley-spec-ledger-test-0.1.0.0-inplace:Test.Shelley.Spec.Ledger.Generator.Trace.Ledger
      Use --quickcheck-replay=567301 to reproduce.
```

File: alonzo/impl/cardano-ledger-alonzo.cabal [hunk 1]
```diff
@@ -14,11 +14,29 @@ copyright:           2020 Input Output (Hong Kong) Ltd.
 category:            Network
 build-type:          Simple
 
-library
+common base                         { build-depends: base                             >= 4.12       && < 4.15     }
```
Comment (nc6) [inline alonzo/impl/cardano-ledger-alonzo.cabal:17]: Why the super-weird formatting on this line?

**Review by nc6 — APPROVED** [review]
Looks good aside from a few weird formatting artifacts.

File: alonzo/impl/cardano-ledger-alonzo.cabal (hunk 1 above)
Comment (newhoggy) [reply alonzo/impl/cardano-ledger-alonzo.cabal:17]: Haha, yeah it is super-weird.

I copied it from here: https://github.com/input-output-hk/cardano-node/blob/import-cardano-submit-api-from-cardano-rest-repo-4/cardano-submit-api/cardano-submit-api.cabal

It's inspired by what Kmett does in his project: https://github.com/ekmett/coda/blob/master/coda.cabal

Which is to be able to specify constraints of any given package exactly once per cabal file.

Of course, I can use more conventional syntax for this stanza too if that's preferable.

Comment (newhoggy) [reply alonzo/impl/cardano-ledger-alonzo.cabal:17]: I've pushed a new commit to use conventional formatting on base stanzas.

**Pull request merged**